### PR TITLE
feat: Add type hints

### PR DESCRIPTION
**Pull request by bradenmacdonald**
This adds type hints to opaque-keys, including to the CI build so that type rules are enforced before PRs can be merged.

This is actually aimed to help with edx-platform development, because once these type hints are in place, mypy can be used to check types of modules in edx-platform - see https://github.com/openedx/edx-platform/pull/32591 for example. Currently mypy just treats any OpaqueKey type as `Any` which is not helpful.

Some of the variables are of type `ObjectId` but this gets broadened automatically to `Any` because we're using an old version of `pymongo` that lacks type annotations. Once we upgrade `pymongo`, we'll have a more specific `ObjectId` type working.

Left for future:
- [ ] Upgrade `pymongo` so we have a working `ObjectId` type

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald!

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.

**Comment by bradenmacdonald** [comment]
@felipemontoya @pdpinch would one of you be able to review this PR for me?

File: opaque_keys/__init__.py
Comment (blarghmatey) [inline opaque_keys/__init__.py:370]: Is it worthwhile (or even possible) to include the subtypes of the tuple returned?

File: mypy.ini
Comment (blarghmatey) [inline mypy.ini:1]: It's probably worth including the [`python_version`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-python_version) to ensure that MyPy is applying the proper typing rules for the target version that this will be running under, rather than just what happens to be running on the machine that executes the mypy command.

File: opaque_keys/__init__.py
Comment (bradenmacdonald) [reply opaque_keys/__init__.py:370]: It would be nice, but I don't think it's even possible, as the number of fields in the tuple depends on the `KEY_FIELDS` of the subclass. The type isn't `tuple[str, ...]` either, as some of the "key field" values are actually `ObjectId` type, not strings.

File: mypy.ini
Comment (bradenmacdonald) [reply mypy.ini:1]: Sure, done.

**Comment by felipemontoya** [comment]
I'm not well versed with the type hints. I don't think I would be a good reviewer but I'm glad to see the PR is being reviewed now.

**Comment by bradenmacdonald** [comment]
@pdpinch would you be able to approve this one?

**Comment by bradenmacdonald** [comment]
Thanks for your review, @blarghmatey :)

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by bradenmacdonald** [comment]
@felipemontoya or @pdpinch would one of you be comfortable approving this based on @blarghmatey's review? Sorry to ping multiple times here, but you two are the only CCs for this repo so I'm not sure how to move it forward otherwise.

**Comment by felipemontoya** [comment]
Yes. I actually learned a lot about type hints while following this PR

**Review by felipemontoya — APPROVED** [review]
Based on the review by @blarghmatey and the inclusion of mypy as part of the test suite this looks good to me.
Thanks @bradenmacdonald for the update to using type hints.

**Comment by openedx-webhooks** [comment]
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/opaque-keys&entry.867055334=https://github.com/openedx/opaque-keys/pull/256&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2023-07-21+17:53&entry.1917517419=2023-08-21+19:51&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by bradenmacdonald** [comment]
Thanks @felipemontoya !

**Pull request merged**
